File: python/happy_number.py
class Solution(object):
    # O(log(n)) time | O(log(n)) space
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        visited = set()
        while n != 1:
            sum = 0
            x = n
            while x > 0:
                digit = x % 10
                x = x // 10
                sum += digit * digit

            if sum in visited:
                return False
            
            visited.add(sum)
            n = sum

        return True
    
solution = Solution()
print(solution.isHappy(19))

File: src/main/java/october_2025/HappyNumber.java
package october_2025;

import java.util.HashSet;
import java.util.Set;

public class HappyNumber {

    public static void main(String[] args) {
        int n = 2;

        HappyNumber happyNumber = new HappyNumber();
        boolean result = happyNumber.isHappy(n);
        System.out.println(result);
    }

    // O(log(n)) time | O(log(n)) space
    public boolean isHappy(int n) {
        Set<Integer> visited = new HashSet<>();
        while (n != 1) {
            int sum = 0;
            int temp = n;

            while (temp > 0) {
                int curr = temp % 10;
                sum += curr * curr;
                temp = temp / 10;
            }
            if (visited.contains(sum)) {
                return false;
            }

            n = sum;
            visited.add(n);
        }
        return true;
    }

}
